pydal/adapters/base.py

```diff
@@ -579,37 +593,28 @@ def _build_joins_for_select(self, tablenames, param):
         ]
         return (
             join_tables, join_on, tables_to_merge, join_on_tables,
-            important_tablenames, excluded
+            important_tablenames, excluded, tablemap
         )
 
     def _select_wcols(self, query, fields, left=False, join=False,
                       distinct=False, orderby=False, groupby=False,
                       having=False, limitby=False, orderby_on_limitby=True,
                       for_update=False, outer_scoped=[], required=None,
                       cache=None, cacheable=None, processor=None):
-        #: parse tablenames
-        tablenames = self.tables(query)
-        tablenames_for_common_filters = tablenames
+        #: parse tablemap
+        tablemap = self.tables(query)
         #: apply common filters if needed
         if use_common_filters(query):
-            query = self.common_filter(query, tablenames_for_common_filters)
-        #: expand query if needed
-        if query:
-            query = self.expand(query)
+            query = self.common_filter(query, list(tablemap.values()))
         #: auto-adjust tables
-        for field in fields:
-            for tablename in self.tables(field):
-                if tablename not in tablenames:
-                    tablenames.append(tablename)
-        if len(tablenames) < 1:
-            raise SyntaxError('Set: no tables selected')
+        tablemap = merge_tablemaps(tablemap, self.tables(*fields))
         #: remove outer scoped tables if needed
-        if outer_scoped:
-            tablenames = [
-                name for name in tablenames if name not in outer_scoped]
-        #: prepare columns and expand fields
-        colnames = list(map(self._colexpand, fields))
-        sql_fields = ', '.join(map(self._geoexpand, fields))
+        for item in outer_scoped:
+            # FIXME: check for name conflicts
```

Comment:
Fixing it would break backward compatibility with third-party code. `outer_scoped` is currently a list of internal table names. It would have to be changed into a dict of table objects, otherwise there is no way to check for conflicts (same name referencing different objects).